docs/tutorial/esm.md

```diff
@@ -0,0 +1,171 @@
+|----------------------|------------|-----------------------|-------------------------|
+| Main                 | Node.js    | N/A                   | <ul><li> [You must use `await` generously before the app's `ready` event](#you-must-use-await-generously-before-the-apps-ready-event) </li></ul> |
+| Renderer (Sandboxed) | Chromium   | Unsupported           | <ul><li> [Sandboxed preload scripts can't use ESM imports](#sandboxed-preload-scripts-cant-use-esm-imports) </li></ul> |
+| Renderer (Unsandboxed & Context Isolated) | Chromium | Node.js | <ul><li> [Unsandboxed ESM preload scripts will run after page load on pages with no content](#unsandboxed-esm-preload-scripts-will-run-after-page-load-on-pages-with-no-content) </li> <li>[ESM Preload Scripts must have the `.mjs` extension](#esm-preload-scripts-must-have-the-mjs-extension)</li></ul> |
+| Renderer (Unsandboxed & Non Context Isolated) | Chromium | Node.js | <ul><li>[Unsandboxed ESM preload scripts will run after page load on pages with no content](#unsandboxed-esm-preload-scripts-will-run-after-page-load-on-pages-with-no-content)</li><li>[ESM Preload Scripts must have the `.mjs` extension](#esm-preload-scripts-must-have-the-mjs-extension)</li><li>[ESM preload scripts must be context isolated to use dynamic Node.js ESM imports](#esm-preload-scripts-must-be-context-isolated-to-use-dynamic-nodejs-esm-imports)</li></ul> |
+
+## Main process
+
+Electron's main process runs in a Node.js context and uses its ESM loader. Usage should follow
+[Node's ESM documentation](https://nodejs.org/api/esm.html). To enable ESM in a file in the
+main process, one of the following conditions must be met:
+
+- The file ends with the `.mjs` extension
+- The nearest parent package.json has `"type": "module"` set
+
+See Node's [Determining Module System](https://nodejs.org/api/packages.html#determining-module-system)
+doc for more details.
+
+### Caveats
+
+#### You must use `await` generously before the app's `ready` event
+
+ES Modules are loaded **asynchronously**. This means that only side effects
+from the main process entry point's imports will execute before the `ready` event.
+
+This is important because certain Electron APIs (e.g. [`app.setPath`](../api/app.md#appsetpathname-path))
+need to be called **before** the app's `ready` event is emitted.
+
+With top-level `await` available in Node ESM, make sure to `await` every Promise that you need to
+execute before the `ready` event. Otherwise, your app may be `ready` before your code executes.
+
+For example, if `index.mjs` calls `import('./set-up-paths.mjs')` at the top level, the app will
+likely already be `ready` by the time that dynamic import resolves.
+
+```js @ts-nocheck title='index.mjs (Main Process)'
+// add an await call here to guarantee that path setup will finish before `ready`
+import('./set-up-paths.mjs')
+
+app.whenReady(() => {
+  console.log('This code may execute before the above import')
+})
```

Comment:
```suggestion
```js @ts-expect-error=[2] title='index.mjs (Main Process)'
// add an await call here to guarantee that path setup will finish before `ready`
import('./set-up-paths.mjs')

app.whenReady().then(() => {
  console.log('This code may execute before the above import')
})
```

Let's prefer `@ts-expect-error` when possible. Otherwise you suppress real errors like in this case.